internal/replaydetector/replaydetector.go

```diff
@@ -0,0 +1,59 @@
+// Package replaydetector provides packet replay detection algorithm.
+package replaydetector
+
+// ReplayDetector is the interface of sequence replay detector.
+type ReplayDetector interface {
+	// Check returns true if given sequence number is not replayed.
+	// Call accept() to mark the packet is received properly.
+	Check(seq uint64) (accept func(), ok bool)
+}
+
+type slidingWindowDetector struct {
+	latestSeq  uint64
+	maxSeq     uint64
+	windowSize uint64
+	mask       []uint64
+}
+
+// New creates ReplayDetector.
+func New(windowSize uint, maxSeq uint64) ReplayDetector {
+	chunkSize := (windowSize + 63) / 64
+	if chunkSize == 0 {
+		chunkSize = 1
+	}
+	return &slidingWindowDetector{
+		windowSize: uint64(windowSize),
+		maxSeq:     maxSeq,
+		mask:       make([]uint64, chunkSize),
+	}
+}
+
+func (d *slidingWindowDetector) Check(seq uint64) (accept func(), ok bool) {
+	if seq+d.windowSize < d.latestSeq || seq > d.maxSeq {
+		// Older than window size or exceeding upper limit.
+		return func() {}, false
+	}
+	if seq > d.latestSeq {
+		// Update the head of the window.
+		shift := seq - d.latestSeq
+		for i := len(d.mask) - 1; i > 0; i-- {
+			d.mask[i] <<= shift
+			d.mask[i] |= d.mask[i-1] >> (64 - shift)
```

Comment:
This loose some bits if window size is larger than 128 and window shift is larger than 64.